### DISCUSSION TITLE
Determine modes or buffers where TAB is ineffective and RET should be used instead

### DISCUSSION THREAD

**Discussion by rswgnu** [magit/magit#4317] [Ideas]
I would like to create or have a key binding in Magit that has the following behavior, mainly for listing modes:  If pressing TAB would have no visible effect, then press RET.

The reason is that I am conditioned to press TAB in many magit modes to expand and collapse sections.  If there is nothing to expand or collapse, then I want to view entries and RET does this but I don't want to press the second key in that case (because I am trying to automate some things).

Is there a way to tell whether TAB is not effective for a particular major mode or to execute the TAB command and determine whether it changed the buffer properties or not? 

Do you have a command suggestion that will do this?

Thanks,

Bob

**Comment by tarsius** [comment]
This should do the trick:

```lisp
(defun my/magit-section-toggle (section)
  ;; My variant of `magit-section-toggle'.  Usage:
  ;; (define-key magit-section-mode-map "TAB" 'my/magit-section-toggle)
  "Toggle visibility of the body of the current section."
  (interactive (list (magit-current-section)))
  (cond ((eq section magit-root-section)
         (user-error "Cannot hide root section"))
        ((or (oref section content)
             (oref section washer))
         (goto-char (oref section start))
         (if (oref section hidden)
             (magit-section-show section)
           (magit-section-hide section)))
        (t
         (if-let ((command (key-binding "RET")))
             (progn (setq last-command-event ?\()
                    (setq this-command command)
                    (call-interactively command))
           (user-error "Nothing to visit either")))))
```

**Comment by rswgnu** [comment]
Excellent, I will give it a try.  Thanks for solving this so quickly.

-- Bob

**Reply by rswgnu** [reply]
It works nicely, just had to add a (kbd) wrapper in this line:

         (if-let ((command (key-binding (kbd "RET"))))
